packages/core/src/database/revert.ts

```diff
@@ -0,0 +1,71 @@
+import { type Checkpoint, encodeCheckpoint } from "@/utils/checkpoint.js";
+import type { HeadlessKysely } from "./kysely.js";
+import type { NamespaceInfo } from "./service.js";
+
+export const revertIndexingTables = async ({
+  checkpoint,
+  namespaceInfo,
+  db,
+}: {
+  namespaceInfo: NamespaceInfo;
+  db: HeadlessKysely<any>;
+  checkpoint: Checkpoint;
+}) => {
+  await db.wrap({ method: "revert" }, async () => {
+    const encodedCheckpoint = encodeCheckpoint(checkpoint);
+
+    await Promise.all(
+      Object.entries(namespaceInfo.internalTableIds).map(
+        async ([tableName, tableId]) => {
+          await db.transaction().execute(async (tx) => {
```

Comment:
I don't think there are any correctness reasons that they all need to be in the same transaction because the operation is isolated to one table. The design right now lets all tables be reverted in parallel.